src/systems/equation_systems.C

```diff
@@ -855,65 +856,104 @@ EquationSystems::build_elemental_solution_vector (std::vector<Number> & soln,
 
   // Localize into 'soln', provided that parallel_soln is not empty.
   // Note: parallel_soln will be empty in the event that none of the
-  // input names were CONSTANT, MONOMIAL variables or there were
-  // simply no CONSTANT, MONOMIAL variables in the EquationSystems
-  // object.
+  // input names were CONSTANT, MONOMIAL nor components of CONSTANT,
+  // MONOMIAL_VEC variables or there were simply none of these in
+  // the EquationSystems object.
   soln.clear();
   if (parallel_soln)
     parallel_soln->localize_to_one(soln);
 }
 
-
-
 std::vector<std::pair<unsigned int, unsigned int>>
 EquationSystems::find_variable_numbers
-  (std::vector<std::string> & names, const FEType * type) const
+  (std::vector<std::string> & names, const FEType * type, const std::vector<FEType> * types) const
 {
   // This function must be run on all processors at once
   parallel_object_only();
 
   libmesh_assert (this->n_systems());
 
-  // If the names vector has entries, we will only populate the soln vector
-  // with names included in that list.  Note: The names vector and var_nums
-  // vector pairs are reordered alphabetically upon exiting this function.
-  std::vector<std::pair<unsigned int, unsigned int>> var_nums;
-  std::vector<std::string> filter_names = names;
-  bool is_filter_names = !filter_names.empty();
-
+  // Resolve class of type input and ensure that at least one of them is null
+  libmesh_error_msg_if(type && types,
+                       "Input either 'type' or 'types' in find_variable_numbers, but not both.");
+
+  std::vector<FEType> type_filter;
+  if (type)
+    type_filter.push_back(*type);
+  else if (types)
+    type_filter = *types;
+
+  // Store a copy of the valid variable names, if any. The names vector will be be repopulated with
+  // any valid names (or all if 'is_names_empty') in the system which pass through the type filter,
+  // if any. If the variable is a vector, its name will be decomposed into its separate components
+  // in accordance with build_variable_names()
+  std::vector<std::string> name_filter = names;
+  bool is_names_empty = name_filter.empty();
   names.clear();
 
+  // initialize convenience variables
+  FEType var_type;
+  std::string name;
+
+  const std::vector<std::string> component_suffix = {"_x", "_y", "_z"};
+  unsigned int dim = _mesh.mesh_dimension();
+  libmesh_error_msg_if(dim > 3, "Invalid dim in find_variable_numbers");
+
+  // Now filter through the variables in each system and store the system index and their index
+  // within that system. This way, we know where to find their data even after we sort them
+  std::vector<std::pair<unsigned int, unsigned int>> var_nums;
+
   for (const auto & pr : _systems)
     {
       const System & system = *(pr.second);
-      const unsigned int nv_sys = system.n_vars();
 
-      for (unsigned int var=0; var < nv_sys; ++var)
+      for (unsigned int var = 0; var < system.n_vars(); ++var)
```

Comment:
Why change this around?  n_vars() isn't changing on us, right?

Reply:
No it's not. I just thought it looked a bit cleaner this way and avoided having to declare a const unsigned int on each system iteration when the only thing we needed it for was setting up the variable loop.

Reply:
That's actually good practice in general - if you declare a local variable, the compiler can just test against it.  E.g. if you don't declare a local variable here, then the compiler can't be sure that e.g. the `System::variable_type()` call later doesn't cause `System::n_vars()` here to change, so it actually has to call `n_vars()` on every loop iteration.

In this particular case it doesn't matter - `n_vars()` is inlinable, and even if it wasn't, that's a huge loop body with a small loop counter and one little function call per counter increment wouldn't make a dent in its performance.  But putting unnecessary-calls-that-the-compiler-can't-tell-are-unnecessary in loop tests is still a good habit to avoid getting into.  We actually have an IntRange class (and some helper functions like `make_range` and `index_range`) that let you get clean looking code without any inefficiency: here you could do `for (auto var : make_range(system.n_vars()))` and it would only evaluate `n_vars()` once.

Reply:
I understand, thanks for the wisdom! I'll go with the `for (auto var : make_range(system.n_vars()))` solution!